osu.Game/Online/Multiplayer/MultiplayerClient.cs

```diff
@@ -243,23 +243,31 @@ public Task ChangeSettings(Optional<string> name = default, Optional<string> pas
         /// <exception cref="InvalidOperationException">If a toggle of ready state is not valid at this time.</exception>
         public async Task ToggleReady()
         {
-            var localUser = LocalUser;
-
-            if (localUser == null)
-                return;
-
-            switch (localUser.State)
+            try
             {
-                case MultiplayerUserState.Idle:
-                    await ChangeState(MultiplayerUserState.Ready).ConfigureAwait(false);
-                    return;
+                var localUser = LocalUser;
 
-                case MultiplayerUserState.Ready:
-                    await ChangeState(MultiplayerUserState.Idle).ConfigureAwait(false);
+                if (localUser == null)
                     return;
 
-                default:
-                    throw new InvalidOperationException($"Cannot toggle ready when in {localUser.State}");
+                switch (localUser.State)
+                {
+                    case MultiplayerUserState.Idle:
+                        await ChangeState(MultiplayerUserState.Ready).ConfigureAwait(false);
+                        return;
+
+                    case MultiplayerUserState.Ready:
+                        await ChangeState(MultiplayerUserState.Idle).ConfigureAwait(false);
+                        return;
+
+                    default:
+                        throw new InvalidOperationException($"Cannot toggle ready when in {localUser.State}");
+                }
+            }
+            catch (NotJoinedRoomException)
+            {
+                Logger.Log("Server reported that we are no longer in the room, exiting locally.");
+                await LeaveRoom().ConfigureAwait(false);
             }
```

Comment:
The log shows the exception to be:
```
2021-08-09 11:14:03 [error]: An unobserved error has occurred.
2021-08-09 11:14:03 [error]: System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (An unexpected error occurred invoking 'ChangeState' on the server. NotJoinedRoomException: This user has not yet joined a multiplayer room.)
2021-08-09 11:14:03 [error]: ---> Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'ChangeState' on the server. NotJoinedRoomException: This user has not yet joined a multiplayer room.
2021-08-09 11:14:03 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2021-08-09 11:14:03 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2021-08-09 11:14:03 [error]: at osu.Game.Online.Multiplayer.MultiplayerClient.ToggleReady()
2021-08-09 11:14:03 [error]: --- End of inner exception stack trace ---
2021-08-09 11:14:03 [error]: An unobserved error has occurred. (inner)
2021-08-09 11:14:03 [error]: Microsoft.AspNetCore.SignalR.HubException: An unexpected error occurred invoking 'ChangeState' on the server. NotJoinedRoomException: This user has not yet joined a multiplayer room.
2021-08-09 11:14:03 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsyncCore(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2021-08-09 11:14:03 [error]: at Microsoft.AspNetCore.SignalR.Client.HubConnection.InvokeCoreAsync(String methodName, Type returnType, Object[] args, CancellationToken cancellationToken)
2021-08-09 11:14:03 [error]: at osu.Game.Online.Multiplayer.MultiplayerClient.ToggleReady()
```

So I think this is supposed to catch `HubException` instead. (can use `when` to catch hub exception when inner is `NotJoinedRoomException`)

Reply:
Pretty sure it's correct as is. The "inner" part is the task throwing an unobserved exception.

Fine to not merge this until someone confirms though; it's quite low priority.

Reply:
I've simulated it by leaving the room right before the `ChangeState` calls, and I can confirm it does not catch the exception by `NotJoinedRoomException`, it needs to be `HubException` to catch it.

And also it appears that the thrown `HubException` has `InnerException` set to null, so there's no actual way to determine the exception is a "not joined room" exception, unfortunately.

Reply:
That seems wrong. The exception types are supposed to be serialised across the line, from what i recall..

Reply:
[This signalr issue thread states otherwise](https://github.com/dotnet/aspnetcore/issues/12633#issuecomment-517035212), unfortunately.

Reply:
I've revisited this one to see what we can do about it. First thought is that if `HubException` is guaranteed to be our responsibility, handling any and all with the same behaviour would be fine. Unfortunately that doesn't seem to be the case, with at least [one internal usage](https://github.com/dotnet/aspnetcore/blob/667f0685b56adc1586464a200766e9b172ad06af/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L296). Maybe it's still fine to always handle and always leave the match?

Thinking this logic could also likely be moved inside `ChangeState` in `OnlineMultiplayerClient`.

Reply:
I'm not sure that handling all errors the same way is going to lead to decent UX. The exceptions raised by the server are sometimes fatal, like in the "left room" case, but sometimes very not much so (e.g. [when a beatmap checksum test fails](https://github.com/ppy/osu-server-spectator/blob/6662a8202411bfc4648d115f2d0eaab17ec437e3/osu.Server.Spectator/Hubs/MultiplayerQueue.cs#L119-L120)).

If that's all going to be turned into a `HubException` client-side with nothing like an inner exception to go off of, then that would mean that receiving this sort of soft error would presumably also kick the user out of the room rather than let them choose the map again which seems a bit... overkill?

Unfortunately it seems that if we want decent non-fatal and fatal server-side error handling to work with signalr we'd have to roll our own by retrofitting a "result" style type into hub operations...

Reply:
I was proposing only for errors from the `ChangeState` flow, to handle what this PR is attempting. But I guess there are some non-fatal ones there (although arguably all errors imply the client is not in a good state, and kicking them from the match once is a good way to recover back).

Reply:
Oh, if it'd only be done on all errors coming out of `ChangeState` then I'd be more inclined to agree with doing so. I think that logic has been there long enough wherein it should work fine most on the time and cause kicks rather rarely, I think?

Reply:
Actually I'm not sure any more. For instance, [this](https://github.com/ppy/osu-server-spectator/blob/4bdcfbb01953cc685376689d635927a25b4830d8/osu.Server.Spectator/Hubs/MultiplayerHub.cs#L729) exception should likely be handled but not cause the user to leave the room.